### DISCUSSION TITLE
NTPoly available on conda-forge

### DISCUSSION THREAD

**Discussion by awvwgk** [william-dawson/NTPoly#175] [General]
I recently packaged NTPoly for the conda package manager. It is now available on the [conda-forge channel](https://anaconda.org/conda-forge/ntpoly) for Linux and MacOS. For more details visit the [NTPoly feedstock](https://github.com/conda-forge/ntpoly-feedstock).

**Comment by william-dawson** [comment]
Thank you for this very interesting work.

I wanted to try out using NTPoly as a library from conda, but ran into some trouble. To start, I created a new conda environment:
```
conda create --name ntpoly-from-conda 
conda activate ntpoly-from-conda
conda install -c conda-forge ntpoly  
```
Then I thought I could create a simple driver program `main.f90`:
```
PROGRAM main
  USE ProcessGridModule
  USE MPI
  IMPLICIT NONE
  INTEGER :: provided, ierr

  !! Start Up
  CALL MPI_Init_thread(MPI_THREAD_SERIALIZED, provided, ierr)
  CALL ConstructProcessGrid(MPI_COMM_WORLD)

  !! Shut Down
  CALL DestructProcessGrid
  CALL MPI_Finalize(ierr)

END PROGRAM
```
To compile it, I supposed I needed to install gfortran into the conda environment:
```
conda install -c conda-forge gfortran 
```
But when I compile the code:
```
mpif90 main.f90 -o conda
```
 I get an error (Mac OSX):
```
Undefined symbols for architecture x86_64:
  "___processgridmodule_MOD_constructprocessgrid_onlyslice", referenced from:
      _MAIN__ in ccERnuVa.o
  "___processgridmodule_MOD_destructprocessgrid", referenced from:
      _MAIN__ in ccERnuVa.o
ld: symbol(s) not found for architecture x86_64
collect2: error: ld returned 1 exit status
```
Ubuntu:
```
/home/wddawson/miniconda3/envs/ntpoly-from-conda/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: 
/tmp/ccZvu3jg.o: in function `MAIN__':
main.f90:(.text+0x37): undefined reference to 
`__processgridmodule_MOD_constructprocessgrid_onlyslice'
/home/wddawson/miniconda3/envs/ntpoly-from-conda/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: 
main.f90:(.text+0x41): undefined reference to 
`__processgridmodule_MOD_destructprocessgrid'
collect2: error: ld returned 1 exit status
```
I guess that installing gfortran from conda was the wrong step, and there are incompatibilities. Do you think there is a solution to this issue?

**Reply by awvwgk** [reply]
I think you forgot to link with the library:

```
mpif90 main.f90 -o test -lNTPoly
mpirun -n 2 ./test
```

For using conda-forge's MPI, we will also need conda-forge's GCC to use the compiler wrappers. Since it is compiled with GCC 9.3 it should be upwards compatible up to GCC 11. The MPICH build also allows usage of [any external compatible MPI implementation](https://conda-forge.org/docs/user/tipsandtricks.html#using-external-message-passing-interface-mpi-libraries).

**Reply by william-dawson** [reply]
Haha, that was a silly mistake on my part, thanks! My plan is to write a news report on the NTPoly website about this feature so thanks for the details!

**Reply by awvwgk** [reply]
You are of course welcome to become a maintainer at conda-forge if you like. Most of the process is automated, *e.g.* if you release a new version here there will be an automatic bump at the feedstock repo by a bot, meaning that it does not require much effort and time to maintain a feedstock.

Just open an issue at the feedstock as described here:
https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-add-user-username
